File: libs/src/python/salmon-core/salmon/salmon_decryptor.py
#!/usr/bin/env python3
'''
MIT License

Copyright (c) 2021 Max Kas

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
'''
import math
import threading
from concurrent.futures import ThreadPoolExecutor

from iostream.memory_stream import MemoryStream
from iostream.random_access_stream import RandomAccessStream
from salmon.integrity.salmon_integrity import SalmonIntegrity
from salmon.integrity.salmon_integrity_exception import SalmonIntegrityException
from salmon.iostream.encryption_mode import EncryptionMode
from salmon.iostream.salmon_stream import SalmonStream
from salmon.salmon_generator import SalmonGenerator
from salmon.salmon_header import SalmonHeader
from salmon.salmon_security_exception import SalmonSecurityException

from typeguard import typechecked


@typechecked
class SalmonDecryptor:
    """
     * Utility class that decrypts byte arrays.
    """

    def __init__(self, threads: int | None = None, buffer_size: int | None = None):
        """
         * Instantiate an encryptor with parallel tasks and buffer size.
         *
         * @param threads    The number of threads to use.
         * @param buffer_size The buffer size to use. It is recommended for performance  to use
         *                   a multiple of the chunk size if you enabled integrity
         *                   otherwise a multiple of the AES block size (16 bytes).
        """

        self.__threads: int = 0
        """
         * The number of parallel threads to use.
        """

        self.__executor: ThreadPoolExecutor | None = None
        """
         * Executor for parallel tasks.
        """

        __buffer_size: int = 0
        """
         * The buffer size to use.
        """

        if threads is None or threads <= 0:
            self.__threads = 1
        else:
            self.__threads = threads
            self.__executor = ThreadPoolExecutor(threads)

        if buffer_size is None:
            self.__buffer_size = SalmonIntegrity.DEFAULT_CHUNK_SIZE
        else:
            self.__buffer_size = buffer_size

    def decrypt(self, data: bytearray, key: bytearray, nonce: bytearray | None,
                has_header_data: bool,
                integrity: bool = False, hash_key: bytearray | None = None, chunk_size: int | None = None) -> bytearray:
        """
         * Decrypt a byte array using AES256 based on the provided key and nonce.
         * @param data The input data to be decrypted.
         * @param key The AES key to use for decryption.
         * @param nonce The nonce to use for decryption.
         * @param hasHeaderData The header data.
         * @param integrity Verify hash integrity in the data.
         * @param hash_key The hash key to be used for integrity.
         * @param chunkSize The chunk size.
         * @return The byte array with the decrypted data.
         * @throws IOError Thrown if there is a problem with decoding the array.
         * @throws SalmonSecurityException Thrown if the key and nonce are not provided.
         * @throws IOError
         * @throws SalmonIntegrityException
        """
        if key is None:
            raise SalmonSecurityException("Key is missing")
        if not has_header_data and nonce is None:
            raise SalmonSecurityException("Need to specify a nonce if the file doesn't have a header")

        if integrity:
            chunk_size = SalmonIntegrity.DEFAULT_CHUNK_SIZE if chunk_size is None else chunk_size

        input_stream: MemoryStream = MemoryStream(data)
        header: SalmonHeader
        header_data: bytearray | None = None
        if has_header_data:
            header = SalmonHeader.parse_header_data(input_stream)
            if header.get_chunk_size() > 0:
                integrity = True
            chunk_size = header.get_chunk_size()
            nonce = header.get_nonce()
            header_data = header.get_header_data()

        if nonce is None:
            raise SalmonSecurityException("Nonce is missing")

        real_size: int = SalmonStream.get_actual_size(data, key, nonce,
                                                      EncryptionMode.Decrypt,
                                                      header_data, integrity, chunk_size, hash_key)
        out_data: bytearray = bytearray(real_size)

        if self.__threads == 1:
            self.__decrypt_data(input_stream, 0, input_stream.length(), out_data,
                                key, nonce, header_data, integrity, hash_key, chunk_size)
        else:
            self.__decrypt_data_parallel(data, out_data,
                                         key, hash_key, nonce, header_data,
                                         chunk_size, integrity)
        return out_data

    def __decrypt_data_parallel(self, data: bytearray, out_data: bytearray,
                                key: bytearray, hash_key: bytearray | None, nonce: bytearray,
                                header_data: bytearray | None,
                                chunk_size: int | None, integrity: bool):
        """
         * Decrypt stream using parallel threads.
         * @param data The input data to be decrypted
         * @param outData The output buffer with the decrypted data.
         * @param key The AES key.
         * @param hash_key The hash key.
         * @param nonce The nonce to be used for decryption.
         * @param headerData The header data.
         * @param chunkSize The chunk size.
         * @param integrity True to verify integrity.
        """

        running_threads: int = 1
        part_size: int = len(data)

        # if we want to check integrity we align to the chunk size otherwise to the AES Block
        min_part_size: int = SalmonGenerator.BLOCK_SIZE
        if integrity and chunk_size is not None:
            min_part_size = chunk_size
        elif integrity:
            min_part_size = SalmonIntegrity.DEFAULT_CHUNK_SIZE

        if part_size > min_part_size:
            part_size = math.ceil(part_size / float(self.__threads))
            # if we want to check integrity we align to the chunk size instead of the AES Block
            rem: int = part_size % min_part_size
            if rem != 0:
                part_size += min_part_size - rem
            running_threads = len(data) // part_size
        else:
            running_threads = 1

        self.__submit_decrypt_jobs(running_threads, part_size,
                                   data, out_data,
                                   key, hash_key, nonce, header_data,
                                   integrity, chunk_size)

    def __submit_decrypt_jobs(self, running_threads: int, part_size: int,
                              data: bytearray, out_data: bytearray,
                              key: bytearray, hash_key: bytearray | None, nonce: bytearray,
                              header_data: bytearray | None,
                              integrity: bool, chunk_size: int | None):

        """
         * Submit decryption parallel jobs.
         * @param runningThreads The number of threads to submit.
         * @param partSize The data length of each part that belongs to each thread.
         * @param data The buffer of data you want to decrypt. This is a shared byte array across all threads where each
         *             thread will read each own part.
         * @param outData The buffer of data containing the decrypted data.
         * @param key The AES key.
         * @param hash_key The hash key for integrity validation.
         * @param nonce The nonce for the data.
         * @param headerData The header data common to all parts.
         * @param integrity True to verify the data integrity.
         * @param chunkSize The chunk size.
        """
        done: threading.Barrier = threading.Barrier(running_threads + 1)
        ex: Exception | None = None
        for i in range(0, running_threads):
            index: int = i

            def decrypt():
                nonlocal ex, index

                try:
                    start: int = part_size * index
                    length: int
                    if index == running_threads - 1:
                        length = len(data) - start
                    else:
                        length = part_size
                    ins: MemoryStream = MemoryStream(data)
                    self.__decrypt_data(ins, start, length, out_data, key, nonce, header_data,
                                        integrity, hash_key, chunk_size)
                except Exception as ex1:
                    ex = ex1
                done.wait()

            self.__executor.submit(lambda: decrypt())

        try:
            done.wait()
        except InterruptedError as ignored:
            pass

        if ex is not None:
            try:
                raise ex
            except Exception as e:
                raise RuntimeError() from e

    def __decrypt_data(self, input_stream: RandomAccessStream, start: int, count: int, out_data: bytearray,
                       key: bytearray, nonce: bytearray,
                       header_data: bytearray | None, integrity: bool, hash_key: bytearray | None,
                       chunk_size: int | None):
        """
         * Decrypt the data stream.
         * @param inputStream The Stream to be decrypted.
         * @param start The start position of the stream to be decrypted.
         * @param count The number of bytes to be decrypted.
         * @param outData The buffer with the decrypted data.
         * @param key The AES key to be used.
         * @param nonce The nonce to be used.
         * @param headerData The header data to be used.
         * @param integrity True to verify integrity.
         * @param hash_key The hash key to be used for integrity verification.
         * @param chunkSize The chunk size.
         * @throws IOError  Thrown if there is an error with the stream.
         * @throws SalmonSecurityException Thrown if there is a security exception with the stream.
         * @throws SalmonIntegrityException Thrown if the stream is corrupt or tampered with.
        """

        stream: SalmonStream | None = None
        output_stream: MemoryStream | None = None
        try:
            output_stream = MemoryStream(out_data)
            output_stream.set_position(start)
            stream = SalmonStream(key, nonce, EncryptionMode.Decrypt, input_stream,
                                  header_data, integrity, chunk_size, hash_key)
            stream.set_position(start)
            total_chunk_bytes_read: int = 0
            # align to the chunk size if available
            buff_size: int = max(self.__buffer_size, stream.get_chunk_size())
            buff: bytearray = bytearray(buff_size)
            bytes_read: int
            while (bytes_read := stream.read(buff, 0, min(len(buff), (
                    count - total_chunk_bytes_read)))) > 0 and total_chunk_bytes_read < count:
                output_stream.write(buff, 0, bytes_read)
                total_chunk_bytes_read += bytes_read
            output_stream.flush()
        except (IOError, SalmonSecurityException, SalmonIntegrityException) as ex:
            print(ex)
            raise SalmonSecurityException("Could not decrypt data") from ex
        finally:
            if input_stream is not None:
                input_stream.close()
            if stream is not None:
                stream.close()
            if output_stream is not None:
                output_stream.close()

    def __del__(self):
        pass
        if self.__executor is not None:
            self.__executor.shutdown(False)

File: libs/src/typescript/salmon-core/salmon/salmon_decryptor.ts
/*
MIT License

Copyright (c) 2021 Max Kas

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/

import { MemoryStream } from "../streams/memory_stream.js";
import { SalmonIntegrity } from "./integrity/salmon_integrity.js";
import { EncryptionMode } from "./streams/encryption_mode.js";
import { SalmonStream } from "./streams/salmon_stream.js";
import { SalmonHeader } from "./salmon_header.js";
import { SalmonSecurityException } from "./salmon_security_exception.js";
import { SalmonAES256CTRTransformer } from "./transform/salmon_aes256_ctr_transformer.js";
import { decryptData } from "./salmon_decryptor_helper.js";

/**
 * Utility class that decrypts byte arrays.
 * Make sure you use setWorkerPath() with the correct worker script.
 */
export class SalmonDecryptor {
    #workerPath = './lib/salmon-core/salmon/salmon_decryptor_worker.js';

    /**
     * The number of parallel threads to use.
     */
    readonly #threads: number;

    /**
     * The buffer size to use.
     */
    readonly #bufferSize: number;

    #promises: Promise<any>[] = [];
    #workers: any[] = [];

    /**
     * Instantiate a decryptor with parallel tasks and buffer size.
     *
     * @param {number} threads The number of threads to use.
     * @param {number} bufferSize The buffer size to use. It is recommended for performance  to use
     *                   a multiple of the chunk size if you enabled integrity
     *                   otherwise a multiple of the AES block size (16 bytes).
     */
    public constructor(threads: number = 0, bufferSize: number = 0) {
        if (threads == 0) {
            this.#threads = 1;
        } else {
            this.#threads = threads;
        }
        if (bufferSize == 0) {
            // we use the chunks size as default this keeps buffers aligned in case
            // integrity is enabled.
            this.#bufferSize = SalmonIntegrity.DEFAULT_CHUNK_SIZE;
        } else {
            this.#bufferSize = bufferSize;
        }
    }

    /**
     * Decrypt a byte array using AES256 based on the provided key and nonce.
     * @param {Uint8Array} data The input data to be decrypted.
     * @param {Uint8Array} key The AES key to use for decryption.
     * @param {Uint8Array | null} nonce The nonce to use for decryption.
     * @param {boolean} hasHeaderData The header data.
     * @param {boolean} integrity Verify hash integrity in the data.
     * @param {Uint8Array | null} hashKey The hash key to be used for integrity.
     * @param {number | null} chunkSize The chunk size.
     * @return {Promise<Uint8Array>} The byte array with the decrypted data.
     * @ Thrown if there is a problem with decoding the array.
     * @throws SalmonSecurityException Thrown if the key and nonce are not provided.
     * @throws IOException Thrown if there is an IO error.
     * @throws IntegrityException Thrown if the data are corrupt or tampered with.
     */
    public async decrypt(data: Uint8Array, key: Uint8Array, nonce: Uint8Array | null,
        hasHeaderData: boolean,
        integrity: boolean = false, hashKey: Uint8Array | null = null, chunkSize: number | null = null): Promise<Uint8Array> {
        if (key == null)
            throw new SalmonSecurityException("Key is missing");
        if (!hasHeaderData && nonce == null)
            throw new SalmonSecurityException("Need to specify a nonce if the file doesn't have a header");

        if (integrity)
            chunkSize = chunkSize == null ? SalmonIntegrity.DEFAULT_CHUNK_SIZE : chunkSize;

        let inputStream: MemoryStream = new MemoryStream(data);
        let header: SalmonHeader | null = null;
        let headerData: Uint8Array | null = null;
        if (hasHeaderData) {
            header = await SalmonHeader.parseHeaderData(inputStream);
            if (header.getChunkSize() > 0)
                integrity = true;
            chunkSize = header.getChunkSize();
            nonce = header.getNonce();
            headerData = header.getHeaderData();
        }
        if (nonce == null)
            throw new SalmonSecurityException("Nonce is missing");

        let realSize: number = await SalmonStream.getActualSize(data, key, nonce, EncryptionMode.Decrypt,
            headerData, integrity, chunkSize, hashKey);
        let outData: Uint8Array = new Uint8Array(realSize);

        if (this.#threads == 1) {
            await decryptData(data, 0, data.length, outData,
                key, nonce, headerData, integrity, hashKey, chunkSize, this.#bufferSize);
        } else {
            await this.#decryptDataParallel(data, outData,
                key, hashKey, nonce, headerData,
                chunkSize, integrity);
        }

        return outData;
    }

    /**
     * Decrypt stream using parallel threads.
     * @param {Uint8Array} data The input data to be decrypted
     * @param outData The output buffer with the decrypted data.
     * @param key The AES key.
     * @param hashKey The hash key.
     * @param nonce The nonce to be used for decryption.
     * @param headerData The header data.
     * @param chunkSize The chunk size.
     * @param integrity True to verify integrity.
     */
    async #decryptDataParallel(data: Uint8Array, outData: Uint8Array,
        key: Uint8Array, hashKey: Uint8Array | null, nonce: Uint8Array, headerData: Uint8Array | null,
        chunkSize: number | null, integrity: boolean): Promise<void> {
        let runningThreads: number = 1;
        let partSize: number = data.length;

        // if we want to check integrity we align to the chunk size otherwise to the AES Block
        let minPartSize: number = SalmonAES256CTRTransformer.BLOCK_SIZE;
        if (integrity && chunkSize != null)
            minPartSize = chunkSize;
        else if (integrity)
            minPartSize = SalmonIntegrity.DEFAULT_CHUNK_SIZE;

        if (partSize > minPartSize) {
            partSize = Math.ceil(data.length / this.#threads);
            if (partSize > minPartSize)
                partSize -= partSize % minPartSize;
            else
                partSize = minPartSize;
            runningThreads = Math.floor(data.length / partSize);
        }

        await this.#submitDecryptJobs(runningThreads, partSize,
            data, outData,
            key, hashKey, nonce, headerData,
            integrity, chunkSize);
    }

    /**
     * Submit decryption parallel jobs.
     * @param runningThreads The number of threads to submit.
     * @param partSize The data length of each part that belongs to each thread.
     * @param data The buffer of data you want to decrypt. This is a shared byte array across all threads where each
     *             thread will read each own part.
     * @param outData The buffer of data containing the decrypted data.
     * @param key The AES key.
     * @param hashKey The hash key for integrity validation.
     * @param nonce The nonce for the data.
     * @param headerData The header data common to all parts.
     * @param integrity True to verify the data integrity.
     * @param chunkSize The chunk size.
     */
    async #submitDecryptJobs(runningThreads: number, partSize: number, data: Uint8Array, outData: Uint8Array,
        key: Uint8Array, hashKey: Uint8Array | null, nonce: Uint8Array,
        headerData: Uint8Array | null, integrity: boolean, chunkSize: number | null): Promise<void> {
        this.#promises = [];
        for (let i = 0; i < runningThreads; i++) {
            this.#promises.push(new Promise(async (resolve, reject) => {
                if (typeof process !== 'object') {
                    if (this.#workers[i] == null)
                        this.#workers[i] = new Worker(this.#workerPath, { type: 'module' });
                    this.#workers[i].removeEventListener('error', null);
				    this.#workers[i].removeEventListener('message', null);
                    this.#workers[i].addEventListener('message', (event: { data: unknown }) => {
                        if(event.data instanceof Error)
							reject(event.data);
						else
							resolve(event.data);
                    });
                    this.#workers[i].addEventListener('error', (event: any) => {
                        reject(event);
                    });
                } else {
                    const { Worker } = await import("worker_threads");
                    if (this.#workers[i] == null)
                        this.#workers[i] = new Worker(this.#workerPath);
                    this.#workers[i].removeAllListeners();
                    this.#workers[i].on('message', (event: any) => {
                        if(event instanceof Error)
							reject(event);
						else
							resolve(event);
                    });
                    this.#workers[i].on('error', (event: any) => {
                        reject(event);
                    });
                }

                let start: number = partSize * i;
                let length: number;
                if (i == runningThreads - 1)
                    length = data.length - start;
                else
                    length = partSize;

                this.#workers[i].postMessage({
                    index: i, data: data, out_size: outData.length, start: start, length: length, key: key, nonce: nonce,
                    headerData: headerData, integrity: integrity, hashKey: hashKey, chunkSize: chunkSize, bufferSize: this.#bufferSize
                });
            }));
        }
        await Promise.all(this.#promises).then((results: any) => {
            for (let i = 0; i < results.length; i++) {
                for (let j = results[i].startPos; j < results[i].endPos; j++) {
                    outData[j] = results[i].outData[j];
                }
            }
        }).catch((event) => {
			console.error(event);
			if(event instanceof Error) {
				throw event;
			} else {
				throw new Error("Could not run Worker, make sure you set the correct workerPath");
			}
        });
    }

    /**
     * Close the decryptor and release associated resources
     */
    public close(): void {
        for (let i = 0; i < this.#workers.length; i++) {
            this.#workers[i].terminate();
            this.#workers[i] = null;
        }
        this.#promises = [];
    }

    /**
     * Set the path where the decryptor worker. This needs to be a relative path starting from
     * the root of your main javascript app.
     * @param {string} path The path to the worker javascript.
     */
    public setWorkerPath(path: string) {
        this.#workerPath = path;
    }

    /**
     * Get the current path for the worker javascript.
     * @returns {string} The path to the worker javascript.
     */
    public getWorkerPath(): string {
        return this.#workerPath;
    }
}
